lightning/restore/restore.go

```diff
@@ -304,48 +305,238 @@ outside:
+					tblName:  viewMeta.Name,
+					stmtType: schemaCreateView,
+					stmts:    make([]*schemaStmt, 0, len(stmts)),
+				}
+				for _, sql := range stmts {
+					restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+						sql: sql,
+					})
+				}
+				err = worker.appendJob(restoreSchemaJob)
+				if err != nil {
+					return err
 				}
-				err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, viewsSchema)
+				// we don't support restore views concurrency, cauz it maybe will raise a error
+				err = worker.wait()
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
 
+func (worker *restoreSchemaWorker) doJob() {
+	var logger log.Logger
+	var session checkpoints.Session
+	defer func() {
+		if session != nil {
+			session.Close()
+		}
+	}()
+	for {
+		select {
+		case <-worker.ctx.Done():
+			return
+		case job := <-worker.jobCh:
+			if job == nil {
+				return
+			}
+			var err error
+			if session == nil {
+				session, err = worker.glue.GetSession(worker.ctx)
+				if err != nil {
+					worker.throw(err)
+					return
+				}
+			}
+			if job.stmtType == schemaCreateDatabase {
+				logger = log.With(zap.String("db", job.dbName))
+			} else if job.stmtType == schemaCreateTable {
+				logger = log.With(zap.String("table", common.UniqueTable(job.dbName, job.tblName)))
+			} else if job.stmtType == schemaCreateView {
+				logger = log.With(zap.String("table", common.UniqueTable(job.dbName, job.tblName)))
+			}
+			for _, stmt := range job.stmts {
+				task := logger.Begin(zap.DebugLevel, fmt.Sprintf("execute SQL: %s", stmt.sql))
+				_, err = session.Execute(worker.ctx, stmt.sql)
 				task.End(zap.ErrorLevel, err)
 				if err != nil {
-					return errors.Annotatef(err, "restore view schema %s failed", dbMeta.Name)
+					switch job.stmtType {
+					case schemaCreateDatabase:
+						err = errors.Annotatef(err, "restore database schema %s failed", job.dbName)
+					case schemaCreateTable:
+						err = errors.Annotatef(err, "restore table schema %s failed", job.tblName)
+					case schemaCreateView:
+						err = errors.Annotatef(err, "restore view schema %s failed", job.tblName)
+					}
+					worker.wg.Done()
+					worker.throw(err)
+					return
 				}
 			}
+			worker.wg.Done()
+		}
+	}
+}
 
+func (worker *restoreSchemaWorker) wait() error {
+	select {
+	case err := <-worker.errCh:
+		if err != nil {
+			worker.quit()
+		}
+		return err
+	case <-worker.ctx.Done():
+		return worker.ctx.Err()
+	default:
+		worker.wg.Wait()
+		return nil
+	}
+}
+
+func (worker *restoreSchemaWorker) throw(err error) {
+	if err != nil {
+		select {
+		case <-worker.ctx.Done():
+			err := worker.ctx.Err()
+			if err != nil {
+				worker.errCh <- err
+			}
+		case worker.errCh <- err:
+			worker.quit()
+		}
+	}
+}
+
+func (worker *restoreSchemaWorker) appendJob(job *schemaJob) error {
+	select {
+	case err := <-worker.errCh:
+		return err
+	case <-worker.ctx.Done():
+		return worker.ctx.Err()
+	case worker.jobCh <- job:
+		worker.wg.Add(1)
```

Comment:
add the wait group before sending to channel, because there's a time slice that `wg` didn't protect the job

Reply:
Means `wg.Add` should be called before send job to `jobCh `, replace job sending case to default case?